Vhost parameter added.

sergiorb opened the pull request:
I've added the vhost parameter so you can customize it. I'll appreciate any feedback on this PR. Thanks in advance.

AlvaroVega commented:
Thaks for your contribution @sergiorb !
Could you add some documentation about new parameter IOTA_AMQP_VHOST in https://github.com/telefonicaid/iotagent-json/blob/master/docs/installationguide.md#amqp-binding-configuration?

fgalan commented on CHANGES_NEXT_RELEASE:
```diff
@@ -1,2 +1,3 @@
 Fix: Set 60 seconds for default mqtt keepalive option (#413)
 Add: Support of multimeasure for MQTT and AMQP transport (#462)
+Add: Virtual host configuration added to AMQP transport
```
```suggestion
Add: Virtual host configuration added to AMQP transport (config.amqp.vhost and IOTA_AMQP_VHOST env var)
```

AlvaroVega commented:
LGTM 
A twin PR for iota-ul is expected

AlvaroVega reviewed (approved):
LGTM

sergiorb commented:
Hello @AlvaroVega @fgalan and sorry for the delay. I've updated the branch with some test that needs a specially set-up RabbitMQ to run due to the creation of the virtual host is not possible from the amqplib library. May you check that out? I'll replicate the changes in [Iotagent-ul](https://github.com/telefonicaid/iotagent-ul) repo as soon as you validate this one. Thanks in advance.

fgalan commented on README.md:
```diff
@@ -119,6 +119,39 @@ All the tests are designed to test end-to-end scenarios, and there are some requ
 -   MQTT v5 broker (like mosquitto v1.6.7 server running)
 -   MongoDB v3.x server running
+-   AMQP 0-9-1 server with `foo/bar` vHost created (lile RabbitMQ v3 server running)
+    -   You can set up Mosquitto to run the test as follows:
```
You mean mosquitto or rabbitmq? The docker command uses `rabbitmq:3` ...

sergiorb replied:
:sweat_smile: I mean RabbitMQ, I don't know why I tend to confuse them haha

fgalan replied:
Ok. Thus this suggestion maybe it an acceptable fix:

```suggestion
    -   You can set up RabbitMQ to run the test as follows:
```

fgalan replied:
Fixed in 6b0da23

fgalan commented on test/unit/amqpBinding-test3.js:
```diff
@@ -0,0 +1,223 @@
+/*
```
instead of -test3.js it would be better a more meaningful suffix, according to the kind of test cases included in this file.

fgalan replied:
Maybe test/unit/amqpBinding-vhost-test.js ?

sergiorb replied:
Nice! I just followed what I found there. Name it as you like.

fgalan commented on test/unit/amqpBinding-vhost-test.js.js:
```diff
@@ -0,0 +1,223 @@
+/*
```
test/unit/amqpBinding-vhost-test.js.js -> test/unit/amqpBinding-vhost-test.js

Too much "js" at the end ;)

fgalan commented:
LGTM. Thanks!

We will merge once you can address the following, please:

* Fix the CHANGES_NEXT_RELASE little conflict existing now
* Sync the changes in the twin PR in IOTA-UL (https://github.com/telefonicaid/iotagent-ul/pull/462)

With regards to README.md, maybe we will do some refactor of the documentation of your contribution but with regards this PR is fine.

fgalan commented on CHANGES_NEXT_RELEASE:
```diff
@@ -1,4 +1,7 @@
-- Upgrade NodeJS version from 10 to 12 in Dockerfile due to Node 10 End-of-Life
-- Set Nodejs 12 as minimum version in packages.json (effectively removing Nodev10 from supported versions)
-- Add list of environment variables which can be protected by Docker Secrets
+- Fix: avoid raise mongo alarm when a measure is not maching a group configuration
+- Fix: Set 60 seconds for default mqtt keepalive option (#413)
+- Upgrade NodeJS version from 10 to 12 in Dockerfile due to Node 10 End-of-Life
+- Set Nodejs 12 as minimum version in packages.json (effectively removing Nodev10 from supported versions)
+- Add: Support of multimeasure for MQTT and AMQP transport (#462)
+- Add: virtual host configuration added to AMQP transport (config.amqp.vhost and IOTA_AMQP_VHOST env var)
+- Add list of environment variables which can be protected by Docker Secrets
```
Not sure if the conflict in CHANGES_NEXT_RELEASE has been correctly solved... CHANGES_NEXT_RELEASE should contain the current entries in master (i.e. https://github.com/telefonicaid/iotagent-json/blob/master/CHANGES_NEXT_RELEASE) plus the line associated to this PR (i.e. ` - Add: virtual host configuration added to AMQP transport (config.amqp.vhost and IOTA_AMQP_VHOST env var)`).

sergiorb commented:
Hello @fgalan and thank you for your feedback. I've updated the `CHANGES_NEXT_RELEASE` file as requested and also updated `README.md` with a more concise note about the RabbitMQ configuration. Notice that rabbitmq:management is necessary for the test to pass. 
I also have pushed these changes to the [telefonicaid/iotagent-ul#462](https://github.com/telefonicaid/iotagent-ul/pull/462/) PR.

fgalan commented:
Could you have a look to the tests run in GitActions, please? It seems many of them are failing...

sergiorb commented:
> Could you have a look to the tests run in GitActions, please? It seems many of them are failing...

I've been working on them and I cannot find a way to mount the definitions file as volume without breaking the pipeline. It seems that `Git checkout` step tries to delete all files, being unable to delete de mounted definition. I've opened a [PR](https://github.com/Namoshek/rabbitmq-github-action/pull/1) in [rabbitmq-github-action](https://github.com/Namoshek/rabbitmq-github-action) in order to remove rabbitmq service and add that action as a step before running the test.

On the other hand, I've ran master-branch's tests locally and the fail too, mainly due to ContextBroker issues. Could you take a look on that?

Thanks.

fgalan commented:
> On the other hand, I've ran master-branch's tests locally and they fail too, mainly due to ContextBroker issues. Could you take a look on that?

Nos sure what do you mean... according to github, test are passing ok in master at the present moment:

![imagen](https://user-images.githubusercontent.com/1534240/112856492-212a5b80-90b0-11eb-9cf5-8ed2eeba2484.png)

fgalan commented:
> Nos sure what do you mean... according to github, test are passing ok in master at the present moment:

After careful examination, this is not fully true... The tests were ok in master when I look at it (the screenshot is correct) but I have re-run the tests in master and they have failed. This may be caused due to two reasons:

* Changes not in the IOTA code itself but in a "open dependency" (i.e. dependency in packages.json that points to a branch instead of a version number of commit). At the present moment, I think iotagent-node-lib is the only dependency of this kind
* Changes in GitHub Actions (although I don't figure out which one...)

We are analyzing it.

CC: @AlvaroVega

fgalan commented:
> > Nos sure what do you mean... according to github, test are passing ok in master at the present moment:
> 
> After careful examination, this is not fully true... The tests were ok in master when I look at it (the screenshot is correct) but I have re-run the tests in master and they have failed. This may be caused due to two reasons:
> 
>    * Changes not in the IOTA code itself but in a "open dependency" (i.e. dependency in packages.json that points to a branch instead of a version number of commit). At the present moment, I think iotagent-node-lib is the only dependency of this kind
>    * Changes in GitHub Actions (although I don't figure out which one could be...)
> 
> 
> We are analyzing it.
> 
> CC: @AlvaroVega

At the end, it was due to iotagent-node-lib. Some changes on that library make some test to be "unaligned". It has been already fixed in master (by PR #550).

Thus, could you please update this PR brach con master. After that I guess that all your changes in ci.yml in this PR are not needed and can be reverted. Remember also to remove packages-lock.json from the PR. Thanks!

fgalan commented:
@sromerotech could you explain why this PR has been closed, please? Do you plan to re-create it? Thanks!
